[gNMI_client] Add an option to trigger memory spike on gNMI server.

yozhao101 opened the pull request:
Signed-off-by: Yong Zhao <yozhao@microsoft.com>

- What is the motivation of this PR?
This PR aims to trigger memory spike on gNMI server side without explicitly closing the channels (TCP connections) from gNMI client side.

- How did you do it?
I added an option `--trigger_mem_spike` in this python client script such that the user can specify this option to trigger memory spike.

- How did you test/verify it?
I verify this on the `ptf` docker container binding to the lab device `str-s6000-acs-11`. The command line I used is as following:
`python /gnxi/gnmi_cli_py/py_gnmicli.py -g -t *.*.*.* -p 50051 -m subscribe -x DOCKER_STATS,TEST_STATS -xt STATE_DB -o "ndastreamingservertest" --trigger_mem_spike`

yozhao101 committed:
[trigger_mem_spike] Add an option in CLI such that this gNMI client tool
can trigger memory spike on gNMI server side.

Signed-off-by: Yong Zhao <yozhao@microsoft.com>

qiluo-msft commented on gnmi_cli_py/py_gnmicli.py:
```diff
@@ -403,19 +405,23 @@ def gen_request(paths, opt, prefix):
   try:
       responses = stub.Subscribe(req_iterator, options['timeout'], metadata=metadata)
+      if exit_early:
+         return 
```
>return [](http://example.com/codeflow?start=9&length=6)

weird indentation #Closed

yozhao101 replied:
Initially I followed the indentation style in this file which used mixed 2 white spaces and 4 white spaced as an indentation. I will update all indentation in this file.

qiluo-msft replied:
Sorry for misleading. I am not asking regarding original code, and only ask you focus on your own code, which indent by 3 blanks.

Since this is a critical feature PR, you'd better keep it small. Could you revert back all the non-critical format change?

yozhao101 replied:
Updated.

qiluo-msft commented on gnmi_cli_py/py_gnmicli.py:
```diff
@@ -403,19 +405,23 @@ def gen_request(paths, opt, prefix):
   max_update_count = options["update_count"]
   try:
       responses = stub.Subscribe(req_iterator, options['timeout'], metadata=metadata)
+      if exit_early:
```
>exit_early [](http://example.com/codeflow?start=9&length=10)

The behavior you want to achieve: when there is missing path in the request, response will get some error, then you exit early. If there is no missing path in the request, no need to exit. #Closed

yozhao101 replied:
As comments posted on the below, if user specifies the option `--trigger_mem_spike`, then there is a while loop which continuoulsy creates a new channel, build a stub and then call the streaming RPC `subscribe(...)` method implemented in the gNMI server side.

Since this option `--trigger_mem_spike` just was used to simulate the high memory usage on the gNMI server side, we do not need get the actual response from gNMI server side. If user specifies `--trigger_mem_spike`, then `exit_early = True` and `susbscribe(...)` method will exit directly instead of waiting for the response from gNMI server. That's why the flag `exit_early` was used at here.

qiluo-msft commented on gnmi_cli_py/py_gnmicli.py:
```diff
@@ -465,11 +471,21 @@ def main():
+  exit_early = False
+  if trigger_mem_spike:
+    exit_early = True
+    while True:
```
>True [](http://example.com/codeflow?start=10&length=4)

Is it an infinite loop? #Closed

yozhao101 replied:
The purpose of option `--trigger_mem_spike` is to simulate the behavior of gNMI C# client which created a new channel to resend the query once it got an exception from gNMI server side. Such behavior will cause high memory usage on the gNMI server.

In order to achieve the similar purpose, the option `--trigger_mem_spike` is added. If user specifies this option in the command line, then this python gNMI client tool will use this loop to continuously create a new channel, build a stub and then call the streaming RPC `subscribe(...)` method implemented in the gNMI server side. This will cause extreme high memory usage on gNMI server side (telemetry process).

qiluo-msft commented on gnmi_cli_py/py_gnmicli.py:
```diff
@@ -157,6 +157,8 @@ def _create_parser():
   parser.add_argument('--encoding', default=0, type=int, help='[0=JSON, 1=BYTES, 2=PROTO, 3=ASCII, 4=JSON_IETF]')
   parser.add_argument('--qos', default=0, type=int, help='')
   parser.add_argument('--use_alias', action='store_true', help='use alias')
+  parser.add_argument('--trigger_mem_spike', action='store_true', help='trigger memory spike on gNMI server side'
```
>trigger_mem_spike [](http://example.com/codeflow?start=25&length=17)

argument name: exit_early #Closed

yozhao101 replied:
Please see the above comments.

yozhao101 committed:
[gNMI_client] Align the indentation format, trigger memory spkie on gNMI
server side and exits automatically if gNMI server was unavailable such
as shutdown.

Signed-off-by: Yong Zhao <yozhao@microsoft.com>

yozhao101 committed:
[gNMI_client] Remove whitespace between equal operator of keyword
argument.

Signed-off-by: Yong Zhao <yozhao@microsoft.com>

yozhao101 commented on gnmi_cli_py/py_gnmicli.py:
```diff
@@ -394,117 +423,142 @@ def gen_request(paths, opt, prefix):
+    certs = _open_certs(**kwargs)
+    creds = _build_creds(target, port, get_cert, certs, notls)
+
+    if trigger_mem_spike:
+        while True:
+            try:
+                stub = _create_stub(creds, target, port, host_override)
+                request_iterator = gen_request(paths, args, prefix)
+                subscribe_start(stub, args, request_iterator)
+            except grpc.RpcError as err:
+                if err.code() == grpc.StatusCode.UNAVAILABLE:
+                    print("Receives an exception '{}' indicating gNMI server shuts down and exiting ..."
+                          .format(err.details()))
+                    sys.exit(1)
```
@qiluo-msft The code snippet from 514 ~ 524 will simulate the behavior of gNMI C# client when it was invoked to poll the telemetry server in production devices.

Specifically, in the PR ( https://github.com/Azure/sonic-mgmt/pull/5296) of sonic-mgmt, test script (https://github.com/Azure/sonic-mgmt/blob/457bf74e6f5491405ea979629e22342def8c5efd/tests/telemetry/test_telemetry.py#L281) will invoke this python gNMI client tool with the option `--trigger_mem_spike` to query the `valid/invalid` paths `DOCKER_STATS,TEST_STATS` in `STATE_DB`. Following is the pattern:

        1. telemetry server received initial query subscribe from Client_IP:port1
        2. Invalid db table Path STATE_DB DOCKER_STATS_TABLE,TEST_STATS_TABLE
        3. Client_IP:port1 shutdown
        4. telemetry server received initial query subscribe from Client_IP:port2
        5. Invalid db table Path STATE_DB DOCKER_STATS_TABLE,TEST_STATS_TABLE
        6. Client_IP:port2 shutdown
        7. Repeat from Step 1

In order to shutdown the gNMI client automatically after running sonic-mgmt test script, code from line 521 ~ 524 was added to detect whether telemetry server was available or not. If telemetry container was restarted, then this python gNMI client will catch such exception and exit automatically.

yozhao101 commented:
@qiluo-msft Can you please help me review this PR again?

yozhao101 commented:
@qiluo-msft @lguohan can you help me review this PR please?

qiluo-msft commented on gnmi_cli_py/py_gnmicli.py:
```diff
@@ -60,331 +60,360 @@
+    """
+    if creds:
+        if host_override:
+            print("Secure channel with host_override is being created ...")
```
>print [](http://example.com/codeflow?start=12&length=5)

Is `print` really needed? Please remove if it is only for your own debug purpose. #Closed

qiluo-msft commented on gnmi_cli_py/py_gnmicli.py:
```diff
@@ -60,331 +60,360 @@
+        return
+
+    if get_cert:
+        print('Obtaining certificate from target with port: {}:{}'.format(target, port))
```
>print [](http://example.com/codeflow?start=8&length=5)

The same #Closed

yozhao101 committed:
Revert "[gNMI_client] Remove whitespace between equal operator of keyword"

This reverts commit 64b7b03c42f3eb8ed5cffa7b58f1c0b58e6d8c40.

yozhao101 committed:
Revert "[gNMI_client] Align the indentation format, trigger memory spkie on gNMI"

This reverts commit 30a9db2c944100d8a96df3c057167ee3cdb6cad6.

yozhao101 committed:
[gnmi_client] Continuously create new gRPC channels after receiving
exception indicating invalid table from server side.

Signed-off-by: Yong Zhao <yozhao@microsoft.com>

yozhao101 committed:
[gnmi_client] Align the indentation.

Signed-off-by: Yong Zhao <yozhao@microsoft.com>

qiluo-msft commented on gnmi_cli_py/py_gnmicli.py:
```diff
@@ -435,8 +437,9 @@ def subscribe_start(stub, options, req_iterator):
-      print("grpc.RpcError received:\n%s" %x)
+  except grpc.RpcError as err:
+      print("Received an exception from server side and error message is: '{}'.".format(err))
+      raise err
```
>err [](http://example.com/codeflow?start=12&length=3)

`raise` without `err` is better.
ref: [https://stackoverflow.com/a/45986945/2514803](https://stackoverflow.com/a/45986945/2514803) #Closed

yozhao101 replied:
Updated.

qiluo-msft commented on gnmi_cli_py/py_gnmicli.py:
```diff
@@ -465,11 +468,25 @@ def main():
   certs = _open_certs(**kwargs)
   creds = _build_creds(target, port, get_cert, certs, notls)
+
+  if trigger_mem_spike:
```
>trigger_mem_spike [](http://example.com/codeflow?start=5&length=17)

I am thinking about the naming. For client point of view, this option is just a loop mode of subscribe.

How about the name `infinite_many_connections`, and it could be applied to any `mode`? #Closed

yozhao101 replied:
Updated the option name. Do we want to implement and reproduce this memory spike issue for all modes such as `get`, `set` and `subscribe`?

qiluo-msft replied:
Yes

qiluo-msft replied:
I see you added new code. My original intention: you can directly add a loop onto below `if/elif/elif` block, the new feature is orthometric to all existing modes. The only difference is oneshot vs. repeating.

yozhao101 replied:
Thanks for great suggestions!

yozhao101 committed:
[gnmi_client] Rename the option name of command line.

Signed-off-by: Yong Zhao <yozhao@microsoft.com>

yozhao101 committed:
[reproduce_mem_spike] Reproduces memory spike issue with `get` and `set`
mode.

Signed-off-by: Yong Zhao <yozhao@microsoft.com>

yozhao101 committed:
[reproduce_mem_spike] Simplify the implementation.

Signed-off-by: Yong Zhao <yozhao@microsoft.com>

qiluo-msft commented:
@lguohan Could you help review and merge?

yozhao101 commented:
@lguohan Guohan, can you help me review this PR to see whether we can merge it or not please?

lguohan commented on gnmi_cli_py/py_gnmicli.py:
```diff
@@ -157,6 +157,8 @@ def _create_parser():
   parser.add_argument('--encoding', default=0, type=int, help='[0=JSON, 1=BYTES, 2=PROTO, 3=ASCII, 4=JSON_IETF]')
   parser.add_argument('--qos', default=0, type=int, help='')
   parser.add_argument('--use_alias', action='store_true', help='use alias')
+  parser.add_argument('--create_infinite_connections', action='store_true', help='trigger memory spike on gNMI server side'
```
why always infinite? do we give options for user to specify the maximum connection they want to establish?

lguohan replied:
for example, we can just setup 1000 connection to trigger memory spike. do we always run test with infinite connections, and always cause client to generate exceptions?

yozhao101 replied:
> for example, we can just setup 1000 connection to trigger memory spike. do we always run test with infinite connections, and always cause client to generate exceptions?

The purpose of this option is to simulate the behavior of GWS C# gNMI client to trigger memory spike by creating TCP connections continuously as the following steps:

    1. telemetry server received initial query subscribe from Client_IP:port1
    2. Invalid db table Path STATE_DB DOCKER_STATS_TABLE,TEST_STATS_TABLE
    3. Client_IP:port1 shutdown
    4. telemetry server received initial query subscribe from Client_IP:port2
    5. Invalid db table Path STATE_DB DOCKER_STATS_TABLE,TEST_STATS_TABLE
    6. Client_IP:port2 shutdown
    7. Repeat from Step 1

In sonic-mgmt testing script, the gNMI client will be invoked with this option to create connections continuously with telemetry server side. Monit script will restart telemetry container if it catches the memory spike issue of telemetry. If telemetry server is restarted, gNMI clinet side will receive an notification/exception indicating telemetry server was shut down and then gNMI client side will exit directly.

yozhao101 replied:
> why always infinite? do we give options for user to specify the maximum connection they want to establish?

I think we can give user an option to specify the maximum connections.

yozhao101 committed:
[gNMI_client] Provides an option `--create_connections` such that user
can specify the number of cient connections they want to create.

Signed-off-by: Yong Zhao <yozhao@microsoft.com>

qiluo-msft commented on gnmi_cli_py/py_gnmicli.py:
```diff
@@ -465,46 +470,68 @@ def main():
+    elif create_connections == 0:
+        break
+    elif create_connections < -1:
+        print('''
```
>print [](http://example.com/codeflow?start=8&length=5)

Instead of manual check arg range and print error, you may leverage built-in capability in ArgumentParser. ref: [https://stackoverflow.com/a/25295717/2514803](https://stackoverflow.com/a/25295717/2514803) #Closed

yozhao101 committed:
[gNMI_client] Simplify the implementation.

Signed-off-by: Yong Zhao <yozhao@microsoft.com>

qiluo-msft commented on gnmi_cli_py/py_gnmicli.py:
```diff
@@ -465,46 +470,69 @@ def main():
-    subscribe_start(stub, args, request_iterator)
+
+  if create_connections < -1:
+    print('''
```
>print [](http://example.com/codeflow?start=4&length=5)

print to stderr. #Closed

yozhao101 replied:
Updated.

qiluo-msft commented on gnmi_cli_py/py_gnmicli.py:
```diff
@@ -465,46 +470,69 @@ def main():
+      if err.code() == grpc.StatusCode.UNAVAILABLE:
+        print("Client receives an exception '{}' indicating gNMI server is shut down and Exiting ..."
+              .format(err.details()))
+        sys.exit(2)
```
>2 [](http://example.com/codeflow?start=17&length=1)

Since you use multiple return code to indicate different failure reasons. Better to define constants. #Closed

yozhao101 replied:
Updated.

yozhao101 committed:
[gNMI_client] Print messages to standard error stream.

Signed-off-by: Yong Zhao <yozhao@microsoft.com>

yozhao101 committed:
[gnmi_client] Uses constants to represent different exit codes.

Signed-off-by: Yong Zhao <yozhao@microsoft.com>

yozhao101 commented:
@lguohan Guohan, can you help me review this PR again please?
